build.rs

```diff
@@ -11,6 +17,148 @@ compile_error!("One of the features `pio` or `native` must be selected.");
+            "esp_idf_full_version=\"{}.{}.{}\"",
+            self.major, self.minor, self.patch
+        ))
+        .chain(once(format!(
+            "esp_idf_version=\"{}.{}\"",
+            self.major, self.minor
+        )))
+        .chain(once(format!("esp_idf_major_version=\"{}\"", self.major)))
+        .chain(once(format!("esp_idf_minor_version=\"{}\"", self.minor)))
+        .chain(once(format!("esp_idf_patch_version=\"{}\"", self.patch)))
+    }
+
+    fn grab_const<T>(text: impl AsRef<str>, const_name: impl AsRef<str>, const_type: impl AsRef<str>) -> Result<T> where T: FromStr, T::Err: error::Error + Send + Sync + 'static {
+        let const_name = const_name.as_ref();
+
+        let value = regex::Regex::new(&format!(r"\s+const\s+{}\s*:\s*{}\s*=\s*(\S+)\s*;", const_name, const_type.as_ref()))?
+            .captures(text.as_ref())
+            .ok_or_else(|| anyhow!("Failed to capture constant {}", const_name))?
+            .get(1)
+            .ok_or_else(|| anyhow!("Failed to capture the value of constant {}", const_name))?
+            .as_str()
+            .parse::<T>()?;
+
+        Ok(value)
+    }
+}
+
 fn main() -> anyhow::Result<()> {
-    build_impl::main()
+    let build_output = build_impl::main()?;
+
+    // We need to restrict the kconfig parameters which are turned into rustc cfg items
+    // because otherwise we would be hitting rustc command line restrictions on Windows
+    //
+    // For now, we take all tristate parameters which are set to true, as well as a few
+    // selected string ones, as per below
+    //
+    // This might change in future
+    let kconfig_str_allow = regex::Regex::new(r"IDF_TARGET")?;
+
+    let cfg_args = build::CfgArgs {
+        args: build_output
+            .kconfig_args
+            .filter(|(key, value)| {
+                matches!(value, kconfig::Value::Tristate(kconfig::Tristate::True)) || kconfig_str_allow.is_match(key)
+            })
+            .filter_map(|(key, value)| value.to_rustc_cfg("esp_idf", key))
+            .collect(),
+    };
+
+    let mcu = cfg_args
+        .get("esp_idf_idf_target")
+        .ok_or_else(|| anyhow!("Failed to get IDF_TARGET from kconfig. cfgs:\n{:?}", cfg_args.args))?;
+
+    let manifest_dir = PathBuf::from(env::var("CARGO_MANIFEST_DIR")?);
+
+    let header_file = path_buf![
+        manifest_dir,
+        "src",
+        "include",
+        if mcu == "esp8266" {
+            "esp-8266-rtos-sdk"
+        } else {
+            "esp-idf"
+        },
+        "bindings.h"
+    ];
+
+    cargo::track_file(&header_file);
+
+    let bindings_file = bindgen::run(
+        build_output
+            .bindgen
+            .builder()?
+            .ctypes_prefix("c_types")
+            .header(header_file.try_to_str()?)
+            .blacklist_function("strtold")
+            .blacklist_function("_strtold_r")
+            .clang_args(if mcu == "esp32c3" {
+                vec!["-target", "riscv32"]
+            } else {
+                vec![]
```

Comment:
Shouldn't we add `-target xtensa` here, to prevent that clang generates something incompatible?

Reply:
Some background info:

In theory at least, we should not be obliged to pass anything in terms of `clang` target, as bindgen is supposed to derive the `clang` target from the rust target for which Cargo is **building** (as opposed to the host target). That's how it always had been in the Pio build.

I think I'm passing the `riscv32` target simply because it doesn't work without it, as we are probably hitting [this](https://github.com/rust-lang/rust-bindgen/issues/1555).

Now, shall we pass always `xtensa` as a target - for all Xtensa based ESP chips - given that their instructions are slightly different from each other (as in ESP32 `xtensa` <> ESP32S2 `xtensa` <> ESP32S3 `xtensa`) - I have no idea and seems dangerous. :) But then, bindgen probably does not care about "instructions" anyway (no logic to care about that), as if it did, we should've been passing `riscv32imc` instead.

Also, the `arch` field of all "Xtensa"  targets in the patched Rust compiler is always set to the `xtensa` string.

Oh well, I'll put `xtensa` there. Probably won't break anything.

Reply:
Yeah, I'm not really should what clang needs to know to generate ABI compatible bindings.

> derive the clang target from the rust target for which Cargo is building
Is this true for xtensa though, because xtensa is not supported upstream?

Either way, we should probably add a TODO comment for this and also a short explanation why we pass them currently.

Reply:
Well, looking at [this](https://github.com/rust-lang/rust-bindgen/blob/master/src/lib.rs#L2157) and [this](https://github.com/rust-lang/rust-bindgen/blob/master/src/lib.rs#L2184) it is clear that it will work for `xtensa` even if we don't specify it explicitly.

It is also a bit clearer why it does not work implicitly for `riscv32`. Probably because clang expects `riscv32-*`, but because of the modular architecture of the RiscV architecture, the rust target is actually `riscv32imc-*`, and what clang gets with the above simplistic code is `riscv32imc-*`, which is perhaps confusing it. Or something along those lines.

In fact, I'm surprised that it is enough to pass `riscv32` (or `xtensa` for that matter) instead of a full LLVM target triple. Oh well, don't plan to dig into this any further for now, will document and push.